[subtitles][ass] Fix completely overlapping subtitle overlays

enen92 opened the pull request:
Superseeds https://github.com/xbmc/xbmc/pull/18330

## Description
When libass is asked to render subtitles, it does it sequentially. It means that if two subtitles/ass events belonging to the same ass track (kodi transforms them into 2 overlays) are contained within the same timeframe, a single call to libass render_images will render both overlay pictures. To accomplish this while avoiding multiple repetitions of the same images on screen, kodi turns the subtitle stream sequential, by modifying the previous added overlay stop PTS with the new overlay start PTS. All works well unless the added overlay is **completely** contained within a previous overlay (thus startPTS > oldOne but stopPTS < oldOne). In that case we should not add the overlay since the timeframe is already contained by old event.

The following example illustrates the problem:

```
Dialogue: 0,0:00:00.00,0:00:01.00,Default,,0,0,0,,{\be3}Should show/hide every 2s
Dialogue: 0,0:00:01.00,0:00:09.00,TOP,,0,0,0,,{\be3\blur3\fad(200,200)}Should be on screen from 1s to 9s
Dialogue: 0,0:00:04.00,0:00:06.00,Default,,0,0,0,,{\be3}Should show/hide every 2s

```

After kodi processes the overlay collection this becomes

```
Dialogue: 0,0:00:00.00,0:00:01.00,Default,,0,0,0,,{\be3}Should show/hide every 2s
Dialogue: 0,0:00:01.00,0:00:06.00,TOP,,0,0,0,,{\be3\blur3\fad(200,200)}Should be on screen from 1s to 9s
Dialogue: 0,0:00:04.00,0:00:06.00,Default,,0,0,0,,{\be3}Should show/hide every 2s
```

The top subtitle that should have been displayed on screen from 1 to 9s is marked to end up at 6 seconds. Thus from 6 to "whatever new subtitle entry" timestamp, no other overlay will appear on screen.

Regarding the implementation:
~Note that the only way an overlay can have the "replace" property set to `true` and without the `stopPTS = 0` is exactly any ASS overlay. Thus, this fix should be well contained and should not introduce any regressions.~

## Motivation and Context
Fixes https://github.com/xbmc/xbmc/issues/18298

## How Has This Been Tested?
With the subtitle file provided in the issue report by @Doraemoe. I tested this subtitle added externally (used to be broken) or muxed into a mkv container (where the issue did not reflect previously). Both display correctly now.
Also tested other subtitled media (srt and pgs) and didn't spot any regressions.

@Doraemoe can I ask you to re-test again? This time should be the last :)

## Screenshots (if appropriate):

**Master:**
![MASTER](https://i.imgur.com/lwS9R3e.png "MASTER")

**PR:**
![PR](https://i.imgur.com/kbFxYuD.png "PR")

Note the top overlay showing as supposed.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

enen92 committed:
[subtitles][ass] Fix completely overlapping subtitle overlays

MediaBrasil commented:
@enen92 
I test here using ISO, it´s broken subtitle, the subtitle continues on screen, wait for the next string, something need fix in your code.

enen92 commented:
@mefiabrasil Either report a bug properly by providing details/subtitle track that helps understand your findings or refrain from posting.

Doraemoe commented:
Hi @enen92, I have tried the same pgs subtitle from last time, while normal subtitle works fine, the lyrics part during ending theme is not showing at all, only the 🎵 symbol is on screen (starting at 1:25:00)

master:
![correct](https://user-images.githubusercontent.com/1129569/91530707-ea8c3700-e94e-11ea-8202-ea0bbe2c6796.png)

this branch:
<img width="1440" alt="error" src="https://user-images.githubusercontent.com/1129569/91530733-f7108f80-e94e-11ea-9926-e4355a571c2d.png">

enen92 commented:
@Doraemoe thanks for the feedback, you're right. This needs to be applied only do ASS subs and not to any subtitle marked as "replaced". Please re-test with the latest implementation/commit.
@MediaBrasil please retest your iso file and check if there's any different behaviour in respect to master

phunkyfish commented on xbmc/cores/VideoPlayer/DVDOverlayContainer.cpp:
```diff
@@ -24,6 +24,8 @@ void CDVDOverlayContainer::Add(CDVDOverlay* pOverlay)
 
   CSingleLock lock(*this);
 
+  bool add = true;
```
I would maybe call the var `addToOverlays` instead of `add`. Only because the word `add` has many meanings. But just a suggestion. Entirely up to you.

enen92 replied:
Will do.
Maybe you can also think of a different name for this method. `Add()` implies that the method just adds the overlay into the container. However this method doesn't only add the overlay but also does some sort of preprocessing on the container itself. When trying to understand the code, the logic here got unnoticed for a while because of the name.

phunkyfish replied:
ProcessAndAddOverlayIfValid()?

enen92 replied:
That sounds right. Will adjust whenever @MediaBrasil confirms he did not spot any regression in his iso file.

Doraemoe commented:
@enen92 I have tested on my end and they all work correctly now. Thank you for working on this 👍

MediaBrasil commented:
All fine now, thanks.

enen92 committed:
[subtitles] Rename CDVDOverlayContainer::Add to CDVDOverlayContainer::ProcessAndAddOverlayIfValid

enen92 commented:
Added a second commit to rename `CDVDOverlayContainer::Add` to `CDVDOverlayContainer:ProcessAndAddOverlayIfValid` as discussed. Also changed the `add` var name to `addToOverlays`. @phunkyfish back to you :)

phunkyfish reviewed (approved):
👍
